osrf/buoy_msgs
# Replicate pbcmd

The physical buoy has a cli called `pbcmd` with several aliases for commands:

available commands aliases for `pbcmd` (help message):
```
pbcmd: Multi-call command Power Buoy dispatcher
Supported commands:
*             bender  - Sets the state of the bender module
*      reset_battery  - Reset battery controller (caution - no args)

*               pump  - Spring Controller pump  off or on for a time in minutes <= 127
*              valve  - Spring Controller valve off or on for a time in seconds <= 127
*             tether  - Spring Controller tether power on or off
*       sc_pack_rate  - Set the CANBUS packet rate from the spring controller
*       reset_spring  - Reset Spring Controller (caution - no args)

*           pc_Scale  - Set the scale factor
*         pc_Retract  - Set the retract factor
*        pc_VTargMax  - Set the max target voltage
*   pc_ChargeCurrLim  - Set the maximum battery charge current
*     pc_DrawCurrLim  - Set the maximum battery current draw
*      pc_BattSwitch  - Set the battery switch state
*            pc_Gain  - Set the gain scheduler gain
*      pc_StdDevTarg  - Set the target RPM standard deviation
*        pc_WindCurr  - Set the winding current target
*        pc_BiasCurr  - Set the winding current bias
*        pc_PackRate  - Set the CANBUS packet rate

*          tf_SetPos  - Open/close the doors in the heave-cone
*    tf_SetActualPos  - Open/close the doors in the heave-cone
*         tf_SetMode  - Set controller mode
*   tf_SetChargeMode  - Set Battery Charge mode
* tf_SetStateMachine  - Set Battery Charge mode
*      tf_SetCurrLim  - Set controller current limit
*        tf_WatchDog  - Toggle controller watchdog (caution - no args)
*           tf_Reset  - Reset Controller (caution - no args)


For help on a command use the command name, i.e., "bender"

Except the reset commands which take no arguments.

DO NOT enter reset_power and expect to get help. The command will execute!
```

For the Sim, only a few of these are implemented:
aliases TODO:
- [x] pump
- [x] valve
- [x] sc_pack_rate
- [x] pc_Scale
- [x] pc_Retract
- [x] pc_WindCurr
- [x] pc_BiasCurr
- [x] pc_PackRate

## Usage:
As part of the build, the post-install step will run the alias installer and install aliases to `~/.local/bin`

## To test:
Build and source the workspace and run the sim, then inspect `~/.local/bin` to see symlinks installed. 

In a new terminal, source the workspace and try running commands:
```
$ pbcmd
```
should print the available commands like above.

try other commands like:
```
$ valve -h
usage: valve [-h] duration_sec

positional arguments:
  duration_sec  Executing valve to Spring Controller valve off or on for a time in seconds <= 12...

optional arguments:
  -h, --help    show this help message and exit
```
should print `valve` usage

missing args should print something like
```
$ valve
usage: valve [-h] duration_sec
valve: error: the following arguments are required: duration_sec
```

and using the command properly:
```
$ valve 10
Executing valve to Spring Controller valve: 10 second(s)
[INFO] [1660956604.401288018] [pbcmd]: /pc_pack_rate_command not available
[INFO] [1660956604.652030356] [pbcmd]: /sc_pack_rate_command not available
[INFO] [1660956604.654093281] [pbcmd]: Command Successful
done!
```
```
[INFO] [1660956615.447699560] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (10s)
[INFO] [1660956623.602440647] [spring_controller]: Valve open (10s)
```
should run the valve in the sim for 10 seconds and:
```
$ valve off
Executing valve to Spring Controller valve: off
[INFO] [1660957253.049844816] [pbcmd]: /pc_pack_rate_command not available
[INFO] [1660957253.300562933] [pbcmd]: /sc_pack_rate_command not available
[INFO] [1660957253.303867220] [pbcmd]: Command Successful
done!
```
```
[INFO] [1660957253.301667417] [spring_controller]: [ROS 2 Spring Control] ValveCommand Received (0s)
[INFO] [1660957510.409397573] [spring_controller]: Valve closed after (2.53s)
```
should turn it off

commands that aren't implemented should let the user know:
```
$ tether
tether not currently implemented
```
